File: 프로그래머스/lv2/43165. 타겟 넘버/타겟 넘버.cpp
#include <queue>
#include <vector>

using namespace std;

int solution(vector<int> numbers, int target) {
    int ans = 0, index = 0;
    queue<int> q;
    q.push(0);
    
    while(index < numbers.size()){
        int size = q.size();
        
        for(int i=0; i<size; i++){
            int now = q.front();
            q.pop();
            
            q.push(now + numbers[index]);
            q.push(now - numbers[index]);
        }
        index++;
    }
    
    while(!q.empty()){
        if(q.front() == target) ans++;
        q.pop();
    }
    
    return ans;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    static int ans = 0;
    
    public void DFS(int sum, int index, int target, int[] numbers) {
        if(index == numbers.length) {
            if(sum == target) ans++;
            return;
        }
        
        DFS(sum + numbers[index], index + 1, target, numbers);
        DFS(sum - numbers[index], index + 1, target, numbers);
    }
    
    public int solution(int[] numbers, int target) {
        DFS(0, 0, target, numbers);
        return ans;
    }
}